File: packages/wasm-api/test/fixtures/string-ptr.ts
// @ts-ignore possibly includes unused imports
import { MemorySlice, Pointer, WasmStringPtr, WasmTypeBase, WasmTypeConstructor } from "@thi.ng/wasm-api";

export interface Foo extends WasmTypeBase {
	single: WasmStringPtr;
	singleMut: WasmStringPtr;
	multi: WasmStringPtr[];
	singlePtr: Pointer<WasmStringPtr>;
	multiPtr: Pointer<WasmStringPtr[]>;
	slice: WasmStringPtr[];
	mutSlice: WasmStringPtr[];
}

export const $Foo: WasmTypeConstructor<Foo> = (mem) => ({
	get align() {
		return 4;
	},
	get size() {
		return 56;
	},
	instance: (base) => {
		let $singlePtr: Pointer<WasmStringPtr> | null = null;
		let $multiPtr: Pointer<WasmStringPtr[]> | null = null;
		let $single: WasmStringPtr | null = null;
		let $singleMut: WasmStringPtr | null = null;
		return {
			get __base() {
				return base;
			},
			get __bytes() {
				return mem.u8.subarray(base, base + 56);
			},
			get single(): WasmStringPtr {
				return $single || ($single = new WasmStringPtr(mem, base, true));
			},
			get singleMut(): WasmStringPtr {
				return $singleMut || ($singleMut = new WasmStringPtr(mem, (base + 8), false));
			},
			get multi(): WasmStringPtr[] {
				const addr = (base + 16);
				const $multi: WasmStringPtr[] = [];
				for(let i = 0; i < 2; i++) $multi.push(new WasmStringPtr(mem, addr + i * 4, true));
				return $multi;
			},
			get singlePtr(): Pointer<WasmStringPtr> {
				return $singlePtr || ($singlePtr = new Pointer<WasmStringPtr>(mem, (base + 32),
				(addr) => new WasmStringPtr(mem, addr, true)
				));
			},
			get multiPtr(): Pointer<WasmStringPtr[]> {
				return $multiPtr || ($multiPtr = new Pointer<WasmStringPtr[]>(mem, (base + 36),
				(addr) => {
					const $buf: WasmStringPtr[] = [];
					for(let i = 0; i < 2; i++) $buf.push(new WasmStringPtr(mem, addr + i * 4, true));
					return $buf;
				}
				));
			},
			get slice(): WasmStringPtr[] {
				const addr = mem.u32[(base + 40) >>> 2];
				const len = mem.u32[(base + 44) >>> 2];
				const $slice: WasmStringPtr[] = [];
				for(let i = 0; i < len; i++) $slice.push(new WasmStringPtr(mem, addr + i * 4, true));
				return $slice;
			},
			get mutSlice(): WasmStringPtr[] {
				const addr = mem.u32[(base + 48) >>> 2];
				const len = mem.u32[(base + 52) >>> 2];
				const $mutSlice: WasmStringPtr[] = [];
				for(let i = 0; i < len; i++) $mutSlice.push(new WasmStringPtr(mem, addr + i * 4, false));
				return $mutSlice;
			},
		};
	}
});

File: packages/wasm-api/test/fixtures/string-ptr.zig
const std = @import("std");
const wasm = @import("wasmapi");

pub const Foo = extern struct {
    single: wasm.ConstStringPtr,
    singleMut: wasm.StringPtr,
    multi: [2]wasm.ConstStringPtr,
    singlePtr: *wasm.ConstStringPtr,
    multiPtr: *[2]wasm.ConstStringPtr,
    slice: wasm.ConstStringPtrSlice,
    mutSlice: wasm.StringPtrSlice,
};

export fn Foo_align() usize {
    return @alignOf(Foo);
}

export fn Foo_size() usize {
    return @sizeOf(Foo);
}

export fn Foo_single_align() usize {
    return @alignOf(wasm.ConstStringPtr);
}

export fn Foo_single_offset() usize {
    return @offsetOf(Foo, "single");
}

export fn Foo_single_size() usize {
    return @sizeOf(wasm.ConstStringPtr);
}

export fn Foo_singleMut_align() usize {
    return @alignOf(wasm.StringPtr);
}

export fn Foo_singleMut_offset() usize {
    return @offsetOf(Foo, "singleMut");
}

export fn Foo_singleMut_size() usize {
    return @sizeOf(wasm.StringPtr);
}

export fn Foo_multi_align() usize {
    return @alignOf([2]wasm.ConstStringPtr);
}

export fn Foo_multi_offset() usize {
    return @offsetOf(Foo, "multi");
}

export fn Foo_multi_size() usize {
    return @sizeOf([2]wasm.ConstStringPtr);
}

export fn Foo_singlePtr_align() usize {
    return @alignOf(*wasm.ConstStringPtr);
}

export fn Foo_singlePtr_offset() usize {
    return @offsetOf(Foo, "singlePtr");
}

export fn Foo_singlePtr_size() usize {
    return @sizeOf(*wasm.ConstStringPtr);
}

export fn Foo_multiPtr_align() usize {
    return @alignOf(*[2]wasm.ConstStringPtr);
}

export fn Foo_multiPtr_offset() usize {
    return @offsetOf(Foo, "multiPtr");
}

export fn Foo_multiPtr_size() usize {
    return @sizeOf(*[2]wasm.ConstStringPtr);
}

export fn Foo_slice_align() usize {
    return @alignOf(wasm.ConstStringPtrSlice);
}

export fn Foo_slice_offset() usize {
    return @offsetOf(Foo, "slice");
}

export fn Foo_slice_size() usize {
    return @sizeOf(wasm.ConstStringPtrSlice);
}

export fn Foo_mutSlice_align() usize {
    return @alignOf(wasm.StringPtrSlice);
}

export fn Foo_mutSlice_offset() usize {
    return @offsetOf(Foo, "mutSlice");
}

export fn Foo_mutSlice_size() usize {
    return @sizeOf(wasm.StringPtrSlice);
}
